src/libraries/Common/src/Interop/Windows/Kernel32/Interop.ReadDirectoryChangesW.cs

```diff
@@ -28,8 +28,8 @@ internal struct FILE_NOTIFY_INFORMATION
             internal FileAction Action;
 
             // Note that the file name is not null terminated
-            private uint FileNameLength;
-            private char _FileName;
+            internal uint FileNameLength;
+            internal char _FileName;
```

Comment:
If you're making this internal you may want a better name than the leading underscore one. Maybe `FileNamePtr` and it's a `char*`  ? Also should these fields be `readonly` ?

However, as written, it would be easy for someone to write code that uses this struct and either forget to validate bounds, or forget to do so before reading `FileName`, as that will AV when it constructs the span, presumably. Would it help if the struct was simply
```c#
        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
        internal struct FILE_NOTIFY_INFORMATION
        {
            internal readonly uint NextEntryOffset;
            internal readonly FileAction Action;

            // Note that the file name is not null terminated
            internal readonly uint FileNameLength;
            internal readonly byte* _FileName;
        }
```

By getting rid of the attractive looking getter that returns a string (but can AV) it's more clear that the caller needs to do validation.

This code below is the only place where FILE_NOTIFY_INFORMATION is used. The `new ReadOnlySpan` stuff would move there.

Reply:
Could you avoid this change by just using span length?

Reply:
>            // Note that the file name is not null terminated
>            internal readonly uint FileNameLength;
>            internal readonly byte* FileNamePtr;

This would not be correct. 

It should be either this:
```
            // Note that the file name is not null terminated
            internal readonly uint FileNameLength;
            internal readonly char FileName;
```
or this: 
```
            // Note that the file name is not null terminated
            internal readonly uint FileNameLength;
            internal fixed char FileName[1];
```

Reply:
> Could you avoid this change by just using span length?

It is best to convert buffer to Span only after the buffer bounds have been validated. Having API that returns Span that points to potentially invalid memory is not a good pattern.

Reply:
I will be using this:
`internal readonly char FileName;`
Because the unit tests passed with it.

I tested this suggestion, but it caused test failures:
`internal fixed char FileName[1];`

<details>
  <summary>Failures</summary>

  ```
  Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. System.ArgumentException: Illegal characters in path. (Parameter 'path')
     at System.IO.Path.GetFullPath(String path) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Path.Windows.cs:line 52
     at System.IO.Tests.FileSystemWatcherTest.<>c__DisplayClass11_0.<WatchRenamed>b__0(Object o, RenamedEventArgs e) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\tests\Utility\FileSystemWatcherTest.cs:line 146
     at System.IO.FileSystemWatcher.NotifyRenameEventArgs(WatcherChangeTypes action, ReadOnlySpan`1 name, ReadOnlySpan`1 oldName) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.cs:line 425
     at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer, UInt32 numBytes) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 317
     at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, AsyncReadState state) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 238
     at System.IO.FileSystemWatcher.<>c.<StartRaisingEvents>b__84_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 63
     at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\ClrThreadPoolBoundHandleOverlapped.cs:line 48
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\Overlapped.cs:line 62
  System.ArgumentException: Illegal characters in path. (Parameter 'path')
     at System.IO.Path.GetFullPath(String path) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Path.Windows.cs:line 52
     at System.IO.Tests.FileSystemWatcherTest.<>c__DisplayClass10_0.<WatchDeleted>b__0(Object o, FileSystemEventArgs e) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\tests\Utility\FileSystemWatcherTest.cs:line 111
     at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, ReadOnlySpan`1 name) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.cs:line 454
     at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer, UInt32 numBytes) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 337
     at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, AsyncReadState state) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 238
     at System.IO.FileSystemWatcher.<>c.<StartRaisingEvents>b__84_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 63
     at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\ClrThreadPoolBoundHandleOverlapped.cs:line 48
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\Overlapped.cs:line 62
  System.ArgumentException: Illegal characters in path. (Parameter 'path')
     at System.IO.Path.GetFullPath(String path) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Path.Windows.cs:line 52
     at System.IO.Tests.FileSystemWatcherTest.<>c__DisplayClass10_0.<WatchDeleted>b__0(Object o, FileSystemEventArgs e) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\tests\Utility\FileSystemWatcherTest.cs:line 111
     at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, ReadOnlySpan`1 name) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.cs:line 454
     at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer, UInt32 numBytes) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 337
     at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, AsyncReadState state) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 238
     at System.IO.FileSystemWatcher.<>c.<StartRaisingEvents>b__84_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 63
     at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\ClrThreadPoolBoundHandleOverlapped.cs:line 48
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\Overlapped.cs:line 62
  System.ArgumentException: Illegal characters in path. (Parameter 'path')
     at System.IO.Path.GetFullPath(String path) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Path.Windows.cs:line 52
     at System.IO.Tests.FileSystemWatcherTest.<>c__DisplayClass9_0.<WatchCreated>b__0(Object o, FileSystemEventArgs e) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\tests\Utility\FileSystemWatcherTest.cs:line 76
     at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, ReadOnlySpan`1 name) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.cs:line 454
     at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer, UInt32 numBytes) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 334
     at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, AsyncReadState state) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 238
     at System.IO.FileSystemWatcher.<>c.<StartRaisingEvents>b__84_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 63
     at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\ClrThreadPoolBoundHandleOverlapped.cs:line 48
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\Overlapped.cs:line 62
  System.ArgumentException: Illegal characters in path. (Parameter 'path')
     at System.IO.Path.GetFullPath(String path) in D:\runtime\src\libraries\System.Private.CoreLib\src\System\IO\Path.Windows.cs:line 52
     at System.IO.Tests.FileSystemWatcherTest.<>c__DisplayClass11_0.<WatchRenamed>b__0(Object o, RenamedEventArgs e) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\tests\Utility\FileSystemWatcherTest.cs:line 146
     at System.IO.FileSystemWatcher.NotifyRenameEventArgs(WatcherChangeTypes action, ReadOnlySpan`1 name, ReadOnlySpan`1 oldName) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.cs:line 425
     at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer, UInt32 numBytes) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 317
     at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, AsyncReadState state) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 238
     at System.IO.FileSystemWatcher.<>c.<StartRaisingEvents>b__84_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer) in D:\runtime\src\libraries\System.IO.FileSystem.Watcher\src\System\IO\FileSystemWatcher.Win32.cs:line 63
     at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\ClrThreadPoolBoundHandleOverlapped.cs:line 48
     at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped) in D:\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Threading\Overlapped.cs:line 62
  ```
</details>